clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -944,47 +944,29 @@ scheduler_tmpl_t<RecordType, ReaderType>::set_initial_schedule(
                     inputs_[input_idx].order_by_timestamp = true;
                 }
             }
-            // Pick the starting inputs by sorting by relative time from each workload's
-            // base_timestamp, which our queue does for us.  We want the rest of the
-            // inputs in the queue in any case so it is simplest to insert all and
-            // remove the first N rather than sorting the first N separately.
-            for (int i = 0; i < static_cast<input_ordinal_t>(inputs_.size()); ++i) {
-                add_to_ready_queue(&inputs_[i]);
-            }
-            for (int i = 0; i < static_cast<output_ordinal_t>(outputs_.size()); ++i) {
-                if (i < static_cast<input_ordinal_t>(inputs_.size())) {
-                    input_info_t *queue_next;
+            // We'll pick the starting inputs below by sorting by relative time from
+            // each workload's base_timestamp, which our queue does for us.
+        }
+        // We need to honor output bindings and possibly time ordering, which our queue
+        // does for us.  We want the rest of the inputs in the queue in any case so it is
+        // simplest to insert all and remove the first N.
+        for (int i = 0; i < static_cast<input_ordinal_t>(inputs_.size()); ++i) {
+            add_to_ready_queue(&inputs_[i]);
+        }
+        for (int i = 0; i < static_cast<output_ordinal_t>(outputs_.size()); ++i) {
+            if (i < static_cast<input_ordinal_t>(inputs_.size())) {
+                input_info_t *queue_next;
 #ifndef NDEBUG
-                    sched_type_t::stream_status_t status =
+                sched_type_t::stream_status_t status =
 #endif
-                        pop_from_ready_queue(i, queue_next);
-                    assert(status == STATUS_OK); // No blocked inputs yet.
-                    if (queue_next == nullptr)
-                        set_cur_input(i, INVALID_INPUT_ORDINAL);
-                    else
-                        set_cur_input(i, queue_next->index);
-                } else
+                    pop_from_ready_queue(i, queue_next);
```

Comment:
I see that the behavior of pop_from_ready_queue depends on the provided `i` (the `for_output` formal param). So is it possible that for some output `i1 < inputs_.size()` we get `queue_next == nullptr`, but then for some  `i2 >= inputs_.size()` we would've got `queue_next != nullptr` had we not entirely skipped this if-block for such an `i2`?

IOW: is it possible that we erroneously leave some input unscheduled even if it could've been scheduled?

Reply:
I think you're right.  The old code for no-time-deps accounted for that but this time-dep code doesn't seem to.  I just removed this final else so it keeps trying to pop from the queue.  I also made a quick unit test for this case: but the test still passes even w/ the short-circuit here, because that output immediately asks for and gets an input before returning any idle record.  So it seems it doesn't really matter.  Even so, I'll leave the fixed/simpler code and the new test.